jade/cli/run_multi_node_job.py

```diff
@@ -0,0 +1,108 @@
+    """Run a job across multiple nodes. The manager node will invoke manager_script_and_args."""
+    output = {}
+    check_run_command(f"jade cluster am-i-manager {jade_runtime_output}", output)
+    result = output["stdout"].strip()
+    if result == "true":
+        ret = run_manager(jade_runtime_output, verbose, manager_script_and_args)
+    else:
+        assert result == "false", result
+        # The only purpose of this worker function is to keep the node allocation
+        # alive. There are more efficient ways of doing this with HPC commands.
+        # However, this procedure allows us to run the JADE JobRunner in the
+        # background on each node and collect resource utilization statistics.
+        ret = run_worker(jade_runtime_output, verbose)
+
+    return ret
+
+
+def run_manager(output_dir, verbose, manager_script_and_args):
+    """Run the manager instance."""
+    shutdown_file = _get_shutdown_file(output_dir)
+    if shutdown_file.exists():
+        os.remove(shutdown_file)
+    try:
+        return _run_manager(output_dir, verbose, manager_script_and_args)
+    finally:
+        # Notify the workers to shutdown.
+        shutdown_file.touch()
+
+
+def _run_manager(output_dir, verbose, manager_script_and_args):
+    filename = os.path.join(output_dir, "run_multi_node_job_manager.log")
+    level = logging.DEBUG if verbose else logging.INFO
+    setup_logging(__name__, filename, file_level=level, console_level=level, mode="w")
+    logger.info("Run manager on %s: %s", socket.gethostname(), get_cli_string())
+
+    # Note that the manager receives its own hostname.
+    output = {}
+    check_run_command(f"jade cluster hostnames {output_dir}", output)
+    hostnames = [x for x in output["stdout"].split() if x != ""]
+    logger.info("Manager found %s hostnames: %s", len(hostnames), hostnames)
+    cmd = " ".join(manager_script_and_args)
+    logger.info("Run manager script [%s]", cmd)
+
+    os.environ["JADE_OUTPUT_DIR"] = output_dir
+    os.environ["JADE_COMPUTE_NODE_NAMES"] = " ".join(hostnames)
+    start = time.time()
+    ret = run_command(cmd)
+    logger.info("Finished job. duration = %s seconds", time.time() - start)
+    return ret
+
+
+def run_worker(output_dir, verbose, poll_interval=60):
+    """Run a worker instance."""
+    hostname = socket.gethostname()
+    filename = os.path.join(output_dir, f"run_multi_node_job_worker_{hostname}.log")
+    level = logging.DEBUG if verbose else logging.INFO
+    setup_logging(__name__, filename, file_level=level, console_level=level, mode="w")
+    logger.info("Run worker: %s", get_cli_string())
+
+    shutdown_file = _get_shutdown_file(output_dir)
+    while not shutdown_file.exists():
+        logger.debug("sleep for %s seconds", poll_interval)
+        time.sleep(poll_interval)
+
+    logger.info("Detected shutdown.")
+    return 0
+
+
+def _get_shutdown_file(output_dir):
+    return Path(output_dir) / SHUTDOWN_FILENAME
```

Comment:
A curiosity: what's the role of shutdown file? looks like it get called in both run_worker and run_manager.

Reply:
The shutdown file is a synchronization mechanism for the manager and workers. The only role of the worker function is to keep the node allocation alive and let the Jade `JobRunner` collect resource statistics.

Once the user script finishes the manager creates the shutdown file. The workers will see that within a minute and then exit. Then all nodes will be returned to the HPC queue.

Reply:
If I have multiple multi node runs, one shutdown file will kill all the nodes in every allocation, right?

Reply:
YES! Thank you.

Reply:
Did you intend to add two underscores in the shutdown file name?

Reply:
Yes. In some configurations job names can have underscores. Whenever I join strings that might have underscores I use double underscore as the delimiter in case I ever need to decode it later.